src/accelerate/accelerator.py

```diff
@@ -2120,7 +2121,17 @@ def gather_for_metrics(self, tensor):
         9
         ```
         """
-        tensor = self.gather(tensor)
+
+        try:
+            recursively_apply(lambda x: x, input_data, error_on_other_type=True)
```

Comment:
Hmm, this seems a bit of a abuse of `recursively_apply` to me (as in: using it for checking the type). It is also a bit wasteful because we recursively traverse the whole data structure here, only to the same again further below when calling `gather`. @muellerzr would it be possibly to to just do something like this instead:

```python
try:
    data = self.gather(input_data)
except TypeError:
    data = gather_object(input_data)
```

i.e. not checking for `all_tensors` at all?

Reply:
gather is a sync op, so doing recursively apply avoids double sync ops and a slowdown on DDP

Reply:
I started with the same strategy as @BenjaminBossan , but then changed to follow other tests "style". So we keep it as it is?